common/functions/src/scalars/udfs/in_basic.rs

```
@@ -15,302 +15,145 @@
-                DataValue::Null => {
-                    continue;
-                }
-                _ => {
-                    return Err(ErrorCode::LogicalError("it's a bug"));
-                }
+    ($T: ident, $INPUT_COL: expr, $ROWS: expr, $COLUMNS: expr, $CAST_TYPE: ident) => {{
+        let mut builder: ColumnBuilder<bool> = ColumnBuilder::with_capacity($ROWS);
+        let mut vals_set = HashSet::with_capacity($ROWS - 1);
+        for col in &$COLUMNS[1..] {
+            let col = cast_column_field(col, &$CAST_TYPE)?;
+            let col_viewer = $T::try_create_viewer(&col)?;
+            if col_viewer.valid_at(0) {
+                let val = col_viewer.value_at(0);
+                vals_set.insert(OrderedFloat::from(val));
             }
         }
-        let arr = $INPUT_ARRAY.$PRIMITIVE_TYPE()?;
-        for val in arr.into_no_null_iter() {
-            let contains = vals_set.contains(val);
-            $BUILDER.push((contains && !NEGATED) || (!contains && NEGATED));
+        let input_viewer = $T::try_create_viewer(&$INPUT_COL)?;
+        for (row, val) in input_viewer.iter().enumerate() {
+            let contains = vals_set.contains(&OrderedFloat::from(val));
+            let valid = input_viewer.valid_at(row);
+            builder.append(valid && ((contains && !NEGATED) || (!contains && NEGATED)));
         }
-    };
+        return Ok(builder.build($ROWS));
+    }};
 }
 
-impl<const NEGATED: bool> Function for InFunction<NEGATED> {
+impl<const NEGATED: bool> Function2 for InFunction<NEGATED> {
     fn name(&self) -> &str {
         "InFunction"
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        let input_dt = args[0].data_type();
-        if input_dt == &DataType::Null {
-            return Ok(DataTypeAndNullable::create(input_dt, false));
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        let input_dt = remove_nullable(args[0]).data_type_id();
```

Here we should check the args type to throw error of unsupported datatype.